src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotUseReferenceEqualsWithValueTypes.cs

```diff
@@ -0,0 +1,98 @@
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+            RuleLevel.BuildWarning,
+            description: s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var objectType = compilationStartContext.Compilation.GetSpecialType(SpecialType.System_Object);
+
+                // Without these symbols the rule cannot run
+                if (objectType == null)
+                {
+                    return;
+                }
+
+                var referenceEqualsMethodGroup = objectType.GetMembers("ReferenceEquals").OfType<IMethodSymbol>();
+                var referenceEqualsMethod = referenceEqualsMethodGroup.GetFirstOrDefaultMemberWithParameterInfos(
+                    GetParameterInfo(objectType),
+                    GetParameterInfo(objectType));
+
+                if (referenceEqualsMethod == null)
+                {
+                    return;
+                }
+
+                compilationStartContext.RegisterOperationAction(operationContext =>
+                {
+                    var invocationExpression = (IInvocationOperation)operationContext.Operation;
+                    var targetMethod = invocationExpression.TargetMethod;
+
+                    if (!referenceEqualsMethod.Equals(targetMethod))
+                    {
+                        return;
+                    }
+
+                    foreach (var argument in invocationExpression.Arguments)
+                    {
+                        var val = argument.Value;
+
+                        if (val is IConversionOperation conversion)
+                        {
+                            val = conversion.Operand;
+                        }
+
+                        if (val.Type.IsValueType)
+                        {
+                            operationContext.ReportDiagnostic(
+                                val.CreateDiagnostic(
+                                    Rule,
+                                    argument.Parameter.Name,
+                                    val.Type.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat)));
+                        }
+                    }
+                },
+                OperationKind.Invocation);
+            });
+        }
+
+        private static ParameterInfo GetParameterInfo(INamedTypeSymbol type, bool isArray = false, int arrayRank = 0, bool isParams = false)
```

Comment:
Consider making the last three parameters of `ParameterInfo.GetParameterInfo` optional and avoiding this private static method.